server/src/remote.rs

```diff
@@ -296,19 +304,236 @@ impl FlightService for FlightServiceHandler {
+
+    // Persist the new model table to the metadata database.
+    let database_path = catalog.data_folder_path.join("metadata.sqlite3");
+    save_model_table_to_database(database_path, &model_table_metadata)
+        .map_err(|error| Status::internal(error.to_string()))?;
+
+    info!("Created model table '{}'.", table_name);
+    catalog.new_model_table_metadata.push(model_table_metadata);
+
+    Ok(())
+}
+
+/// Save the created model table to the database. This includes creating a tags table for the
+/// model table, adding a new row in the model_table_metadata table, and adding a row to the
+/// model_table_columns table for each field column.
+fn save_model_table_to_database(
+    database_path: PathBuf,
+    model_table_metadata: &NewModelTableMetadata,
+) -> Result<(), rusqlite::Error> {
+    // Create a transaction to ensure the database state is consistent across tables.
+    let mut connection = Connection::open(database_path)?;
+    let transaction = connection.transaction()?;
+
+    // Add a column definition for each tag field in the schema.
+    let mut tag_columns = "".to_owned();
+    let mut tag_indices = model_table_metadata.tag_column_indices.iter().peekable();
+
+    while let Some(index) = tag_indices.next() {
+        // If it is the last column in the query, adding a "," results in an SQL syntax error.
+        let suffix = if tag_indices.peek().is_none() {""} else {", "};
+        let field = model_table_metadata.schema.field(*index as usize);
+
+        tag_columns.push_str(format!("{} TEXT NOT NULL{}", field.name(), suffix).as_str());
+    }
+
+    // Create a table_name_tags SQLite table to save the 54-bit tag hashes when ingesting data.
+    // The query is executed with a formatted string since CREATE TABLE cannot take parameters.
+    transaction.execute(format!("CREATE TABLE {}_tags (hash BLOB PRIMARY KEY, {})",
+        model_table_metadata.name, tag_columns).as_str(), ())?;
+
+    // Add a new row in the model_table_metadata table to persist the model table.
+    transaction.execute(
+        "INSERT INTO model_table_metadata (table_name, schema, timestamp_column_index, tag_column_indices)
+             VALUES (?1, ?2, ?3, ?4)",
+        params![
+            model_table_metadata.name,
+            model_table_metadata.schema.to_json().to_string(),
+            model_table_metadata.timestamp_column_index,
+            model_table_metadata.tag_column_indices
+        ]
+    )?;
+
+    // Add a row for each field column to the model_table_columns table.
+    let mut insert_statement = transaction.prepare(
+        "INSERT INTO model_table_columns (table_name, column_name, column_index)
+             VALUES (?1, ?2, ?3)")?;
+
+    for (index, field) in model_table_metadata.schema.fields().iter().enumerate() {
+        // Only add a row for the field if it is not the timestamp or a tag.
+        let in_tag_indices = model_table_metadata.tag_column_indices.contains(&(index as u8));
```

Comment:
Is there a a step missing that converts the `tag_column_indices` from `u8` into `usize` before this comparison or is it a requirements that the [indices of tag columns never overflows a `u8`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=d325f707981029b434b89a28ebd5c5ed)?

Reply:
Currently there is a limitation imposed by the implementation that limits the tag column indices to be at most 255. When we convert the tag column indices bytes in the action body we just assume that each byte is a single index. To fix this issue we either have to change it to assume each index is two bytes or use some kind of delimiter between the tag column indices.

Reply:
It it probably fine for now, but we should probably document using comments and a task.

Reply:
I have added a comment documenting the current assumption and a [work item](https://dev.azure.com/ModelarData/ModelarDB%20Implementation/_workitems/edit/240) to lift the limitation.